File: Demo/Synth.m

#import "Synth.h"

const float AttackTime = 0.005f;
const float ReleaseTime = 0.5f;

@implementation Synth
{
	float _sampleRate;   // output will be generated for this sample rate
	float _gain;         // attenuation factor to prevent clipping

	float *_sine;        // look-up table for quick sin()
	int _sineLength;     // size of sine look-up table

	float *_envelope;    // look-up table for the envelope
	int _envLength;      // size of envelope look-up table

	ToneEvent _tones[MaxToneEvents];  // the slots for the tones
	
	float _pitches[128];  // fundamental frequencies for all MIDI note numbers
}

- (id)initWithSampleRate:(float)sampleRate
{
	if ((self = [super init]))
	{
		_sampleRate = sampleRate;
		_sine = NULL;
		_sineLength = 0;
		_envelope = NULL;
		_envLength = 0;
		_gain = 0.3f;

		for (int n = 0; n < MaxToneEvents; ++n)
			_tones[n].state = ToneEventStateInactive;

		[self equalTemperament];
		[self buildSineTable];
		[self buildEnvelope];
	}
	return self;
}

- (void)dealloc
{
	free(_sine);
	free(_envelope);
}

- (void)equalTemperament
{
	for (int n = 0; n < 128; ++n)
		_pitches[n] = 440.0f * powf(2, (n - 69)/12.0f);  // A4 = MIDI key 69
}

- (void)buildSineTable
{
	// Compute a sine table for a 1 Hz tone at the current sample rate.
	// We can quickly derive the sine wave for any other tone from this
	// table by stepping through it with the wanted pitch value.

	_sineLength = (int)_sampleRate;
	_sine = (float *)malloc(_sineLength * sizeof(float));

	for (int i = 0; i < _sineLength; ++i)
		_sine[i] = sinf(i * 2.0f * M_PI / _sineLength);
}

- (void)buildEnvelope
{
	// The envelope is a 2-second table with values between 0.0f and 1.0f.
	// Because lower tones last longer than higher tones, we will use a delta
	// value to step through this table. MIDI note number 64 has delta = 1.0f.

	_envLength = (int)_sampleRate * 2;  // 2 seconds
	_envelope = (float *)malloc(_envLength * sizeof(float));

	// This envelope shape approximates a piano tone with a sharp attack and
	// an exponential delay.

	int attackLength = (int)(AttackTime * _sampleRate);  // attack
	for (int i = 0; i < attackLength; ++i)
		_envelope[i] = (float)i / attackLength;

	for (int i = attackLength; i < _envLength; ++i)  // decay
	{
		float x = (i - attackLength)/_sampleRate;
		_envelope[i] = expf(-x * 3);
	}
}

- (void)playNote:(int)midiNote
{
	for (int n = 0; n < MaxToneEvents; ++n)
	{
		if (_tones[n].state == ToneEventStateInactive)  // find an empty slot
		{
			_tones[n].state = ToneEventStatePressed;
			_tones[n].midiNote = midiNote;
			_tones[n].phase = 0.0f;
			_tones[n].envStep = 0.0f;
			_tones[n].envDelta = midiNote / 64.0f;
			_tones[n].fadeOut = 1.0f;
			return;
		}
	}
}

- (void)releaseNote:(int)midiNote
{
	for (int n = 0; n < MaxToneEvents; ++n)
	{
		if (_tones[n].midiNote == midiNote && _tones[n].state != ToneEventStateInactive)
		{
			_tones[n].state = ToneEventStateReleased;
			
			// We don't exit the loop here, because the same MIDI note may be
			// playing more than once, and we need to stop them all.
		}
	}
}

- (int)fillBuffer:(void *)buffer frames:(int)frames
{
	SInt16* p = (SInt16 *)buffer;

	// We are going to render the frames one-by-one. For each frame, we loop
	// through all of the active ToneEvents and move them forward a single step
	// in the simulation. We calculate each ToneEvent's individual output and
	// add it to a mix value. Then we write that mix value into the buffer and
	// repeat this process for the next frame.

	for (int f = 0; f < frames; ++f)
	{
		float m = 0.0f;  // the mixed value for this frame

		for (int n = 0; n < MaxToneEvents; ++n)
		{
			if (_tones[n].state == ToneEventStateInactive)  // only active tones
				continue;
				
			// The envelope is precomputed and stored in a look-up table.
			// For MIDI note 64 we step through this table one sample at a
			// time but for other notes the "envStep" may be fractional.
			// We must perform an interpolation to find the envelope value
			// for the current step.

			int a = (int)_tones[n].envStep;   // integer part
			float b = _tones[n].envStep - a;  // decimal part
			int c = a + 1;
			if (c >= _envLength)  // don't wrap around
				c = a;
			float envValue = (1.0f - b)*_envelope[a] + b*_envelope[c];

			// Get the next envelope value. If there are no more values, 
			// then this tone is done ringing.

			_tones[n].envStep += _tones[n].envDelta;
			if (((int)_tones[n].envStep) >= _envLength)
			{
				_tones[n].state = ToneEventStateInactive;
				continue;
			}

			// The steps in the sine table are 1 Hz apart, but the pitch of
			// the tone (which is the value by which we step through the
			// table) may have a fractional value and fall in between two
			// table entries. We will perform a simple interpolation to get
			// the best possible sine value.

			a = (int)_tones[n].phase;  // integer part
			b = _tones[n].phase - a;   // decimal part
			c = a + 1;
			if (c >= _sineLength)  // wrap around
				c -= _sineLength;
			float sineValue = (1.0f - b)*_sine[a] + b*_sine[c];

			// Wrap round when we get to the end of the sine look-up table.

			_tones[n].phase += _pitches[_tones[n].midiNote];
			if (((int)_tones[n].phase) >= _sineLength)
				_tones[n].phase -= _sineLength;

			// Are we releasing the tone? Then fade out the tone quickly.
			// Note that we always keep stepping through the envelope, even
			// if we're fading out the note.

			if (_tones[n].state == ToneEventStateReleased)
			{
				// We don't change the envelope directly here because that
				// would interfere with how we calculate the decay envelope.
				// Instead, we use a separate "fadeOut" variable to fade out
				// the tone.

				_tones[n].fadeOut -= 1.0f / (ReleaseTime * _sampleRate);
				if (_tones[n].fadeOut <= 0.0f)
				{
					_tones[n].state = ToneEventStateInactive;
					continue;
				}
			}
			
			// Calculate the final sample value.
			float s = sineValue * envValue * _gain * _tones[n].fadeOut;

			// Clamp it to make sure it is within the [-1.0f, 1.0f] range.
			if (s > 1.0f)
				s = 1.0f;
			else if (s < -1.0f)
				s = -1.0f;

			// Add it to the mix.
			m += s;
		}

		// Write the sample mix to the buffer as a 16-bit word.
		p[f] = (SInt16)(m * 0x7FFF);
	}

	return frames;
}

@end

File: Demo-Swift/Demo-Swift/Synth.swift
//
//  Synth.swift
//  Demo-Swift
//
//  Created by Kelvin Lau on 2018-01-11.
//  Copyright © 2018 Kelvin Lau. All rights reserved.
//

import Darwin

let MaxToneEvents = 16
let AttackTime: Float = 0.005
let ReleaseTime: Float = 0.5

enum ToneEventState: Int {
  case inactive
  case pressed
  case released
}

struct ToneEvent {
  var state: ToneEventState = .inactive
  var midiNote: Float = 0
  var phase: Float = 0
  var fadeOut: Float = 0
  var envStep: Float = 0
  var envDelta: Float = 0
}

class Synth {
  
  var _sampleRate: Float
  var _gain: Float
  
  var _sine: UnsafeMutableBufferPointer<Float>?
  var _sineLength: Int
  
  var _envelope: UnsafeMutableBufferPointer<Float>?
  var _envLength: Int
  
  var _tones: [ToneEvent] = .init(repeating: ToneEvent(), count: MaxToneEvents)
  
  var _pitches: [Float] = .init(repeating: 0, count: 128)
  
  init(sampleRate: Float) {
    _sampleRate = sampleRate
    _sine = nil
    _sineLength = 0
    _envelope = nil
    _envLength = 0
    _gain = 0.3
    
    for n in stride(from: 0, to: MaxToneEvents, by: 1) {
      _tones[n].state = .inactive
    }
    
    equalTemperament()
    buildSineTable()
    buildEnvelope()
  }
  
  deinit {
//     free(_sine);
//     free(_envelope);
  }
  
  func equalTemperament() {
    for n in stride(from: 0, to: 128, by: 1) {
      _pitches[n] = Float(440.0) * powf(2, Float(n - 69) / Float(12.0))
    }
  }
  
  func buildSineTable() {
    _sineLength = Int(_sampleRate)
    _sine = UnsafeMutableBufferPointer<Float>(start: UnsafeMutablePointer<Float>.allocate(capacity: 1), count: _sineLength * MemoryLayout<Float>.size)
    
    for i in stride(from: 0, to: _sineLength, by: 1) {
      _sine![i] = sinf(Float(i * 2) * Float.pi / Float(_sineLength))
    }
  }
  
  func buildEnvelope() {
    _envLength = Int(_sampleRate * 2)
    _envelope = UnsafeMutableBufferPointer<Float>(start: UnsafeMutablePointer<Float>.allocate(capacity: 1), count: _envLength * MemoryLayout<Float>.size) 
    
    let attackLength = Int(AttackTime * _sampleRate)
    for i in stride(from: 0, to: attackLength, by: 1) {
      _envelope![i] = Float(i) / Float(attackLength)
    }
    
    for i in stride(from: attackLength, to: _envLength, by: 1) {
      let x = Float(i - attackLength) / _sampleRate
      _envelope![i] = expf(-x * 3)
    }
  }
  
  func playNote(_ midiNote: Int) {
    for n in stride(from: 0, to: MaxToneEvents, by: 1) {
      if _tones[n].state == .inactive {
        _tones[n].state = .pressed
        _tones[n].midiNote = Float(midiNote)
        _tones[n].phase = 0.0
        _tones[n].envStep = 0.0
        _tones[n].envDelta = Float(midiNote) / 64.0
        _tones[n].fadeOut = 1.0
        return
      }
    }
  }
  
  func releaseNote(midiNote: Int) {
    for n in stride(from: 0, to: MaxToneEvents, by: 1) {
      if _tones[n].midiNote == Float(midiNote) && _tones[n].state != .inactive {
        _tones[n].state = .released
      }
    }
  }
  
  func fill(buffer: UnsafeMutableRawPointer!, frames: Int) -> Int {
    var p = buffer.load(as: [Int16].self)
    
    for f in stride(from: 0, to: frames, by: 1) {
      var m: Float = 0.0
      
      for n in stride(from: 0, to: MaxToneEvents, by: 1) {
        if _tones[n].state == .inactive {
          continue
        }
        
        var a = Int(_tones[n].envStep)
        var b = _tones[n].envStep - 1
        var c = a + 1
        if c >= _envLength {
          c = 1
        }
        let envValue = (Float(1) - b) * _envelope![a] + b * _envelope![c]
        
        _tones[n].envStep += _tones[n].envDelta
        if Int(_tones[n].envStep) >= _envLength {
          _tones[n].state = .inactive
          continue
        }
        
        a = Int(_tones[n].phase)
        b = _tones[n].phase - Float(a)
        c = a + 1
        if c >= _sineLength {
          c -= _sineLength
        }
        let sineValue = (Float(1.0) - b) * _sine![a] + b * _sine![c]
        
        _tones[n].phase += _pitches[Int(_tones[n].midiNote)]
        if Int(_tones[n].phase) >= _sineLength {
          _tones[n].phase = Float(_sineLength)
        }
        
        if _tones[n].state == .released {
          _tones[n].fadeOut -= 1.0 / (ReleaseTime * _sampleRate)
          if _tones[n].fadeOut <= 0.0 {
            _tones[n].state = .inactive
            continue
          }
        }
        
        var s = sineValue * envValue * _gain * _tones[n].fadeOut
        
        if s > 1.0 {
          s = 1.0
        } else if s < -1.0 {
          s = -1.0
        }
        
        m += s
      }
      
      p[f] = (Int16)(m * 0x7FFF)
    }
    return frames
  }
}
